src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -563,65 +794,159 @@ class segment_meta_cstore::impl
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.lower_bound(o()));
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     upper_bound(model::offset o) const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.upper_bound(o()));
     }
 
     std::optional<segment_meta> last_segment() const {
+        if (!_write_buffer.empty()) {
+            auto last_committed_offset = _col.last_committed_offset().value_or(
+              model::offset::min());
+            auto buffer_last_base_offset = _write_buffer.crbegin()->first;
+            // CHECK: base_offset is never equal to committed_offset
+            if (likely(buffer_last_base_offset > last_committed_offset)) {
+                // return last one in _write_buffer
+                return _write_buffer.crbegin()->second;
+            }
+            // fallthrough
+        }
         return _col.last_segment();
     }
 
-    void insert(const segment_meta& m) { _col.append(m); }
+    void insert(const segment_meta& m) {
+        auto [m_it, insertion_took_place] = _write_buffer.insert_or_assign(
+          m.base_offset, m);
+        // m could be a replacement segment for one in the buffer in 2 cases:
+        // 1. !insertion_took_place (assignment):
+        //    - key was already in use, m is a replacement for a segment in the
+        //    buffer
+        // 2. insertion took place at the start of the buffer:
+        //    - m could be a replacement for a segment already in the store, and
+        //    there could be an overlap with the segments at the start of the
+        //    write buffer
+        // TODO could it happen that a segment does not replace an existing one
+        // (different base_offset), but still overlaps with the successors? i.e.
+        // a gap in the data where a previous segment is lost, and the current
+        // one would be a replacement for the previous one?
+        if(_write_buffer.size()>1 &&
+            (!insertion_took_place || // case 1
+             m_it==_write_buffer.begin()) // case 2
+            ){
+            //
+            auto not_replaced_segment = std::find_if(
+              std::next(m_it), _write_buffer.end(), [&](auto& kv) {
+                  return kv.first > m.committed_offset;
+              });
+            // if(next(m_it) == not_replaced_segment) there is nothing to erase,
+            // _write_buffer.erase would do nothing
+            _write_buffer.erase(std::next(m_it), not_replaced_segment);
+        }
+
+        if (_write_buffer.size() > max_buffer_entries) {
+            flush_write_buffer();
+        }
+    }
 
-    auto inflated_actual_size() const { return _col.inflated_actual_size(); }
+    auto inflated_actual_size() const {
+        // TODO how to deal with write_buffer? is it ok to return an approx
+        // value by not flushing?
```

Comment:
This holds true only if we keep a vector of replaced segments as a member, right? otherwise, without performing the merge operation, it's not possible to know exactly this value (not that this needs to be exact, i think)